pkg/executor/sortexec/parallel_sort_worker.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sortexec
+
+import (
+	"sync"
+
+	"github.com/pingcap/tidb/pkg/util/chunk"
+	"github.com/pingcap/tidb/pkg/util/memory"
+	"golang.org/x/exp/slices"
+)
+
+// SignalCheckpointForSort indicates the times of row comparation that a signal detection will be triggered.
+const SignalCheckpointForSort uint = 20000
+
+type parallelSortWorker struct {
+	lessRowFunc func(chunk.Row, chunk.Row) int
+
+	globalSortedRowsQueue *sortedRowsList
+
+	waitGroup *sync.WaitGroup
+	result    *sortedRows
+
+	mpmcQueue    *chunk.MPMCQueue
+	checkError   func() error
+	processError func(error)
+
+	totalMemoryUsage  int64
+	timesOfRowCompare uint
+
+	memTracker *memory.Tracker
+
+	// Temporarily store rows that will be sorted.
+	rowBuffer sortedRows
+}
+
+func newParallelSortWorker(
+	lessRowFunc func(chunk.Row, chunk.Row) int,
+	globalSortedRowsQueue *sortedRowsList,
+	waitGroup *sync.WaitGroup,
+	result *sortedRows,
+	mpmcQueue *chunk.MPMCQueue,
+	checkError func() error,
+	processError func(error),
+	memTracker *memory.Tracker) *parallelSortWorker {
+	return &parallelSortWorker{
+		lessRowFunc:           lessRowFunc,
+		globalSortedRowsQueue: globalSortedRowsQueue,
+		waitGroup:             waitGroup,
+		result:                result,
+		mpmcQueue:             mpmcQueue,
+		checkError:            checkError,
+		processError:          processError,
+		totalMemoryUsage:      0,
+		timesOfRowCompare:     0,
+		memTracker:            memTracker,
+	}
+}
+
+// Fetching chunks from MPMCQueue and sort them.
+// Rows are sorted only inside a chunk.
+// Return false if there is some error.
+func (p *parallelSortWorker) fetchChunksAndSort() bool {
+	for {
+		err := p.checkError()
+		if err != nil {
+			return false
+		}
+
+		// Memory usage of the chunk has been tracked at the producer side.
+		chk, res := p.mpmcQueue.Pop()
+		if res != chunk.OK {
+			// Check if the queue is closed by an error.
+			err := p.checkError()
+			return err == nil
+		}
+
+		p.totalMemoryUsage += chk.MemoryUsage
+
+		sortedRows := p.sortChunkAndGetSortedRows(chk.Chk)
+		p.globalSortedRowsQueue.add(sortedRows)
+	}
+}
+
+func (p *parallelSortWorker) keyColumnsLess(i, j chunk.Row) int {
+	if p.timesOfRowCompare >= SignalCheckpointForSort {
+		// Trigger Consume for checking the NeedKill signal
+		p.memTracker.Consume(1)
```

Comment:
some hacks...
Wouldn't `memory_tracker.consume` be triggered when merging two rows?